### DISCUSSION TITLE
[ARCHIVED] Обсуждение первого показа новой структуры команде

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#18] [Archive]
## Итоги
- models => entities
- processes => controllers?
- models есть организмы?
- models - могут вызываться и не до фич
- graphql : queries.tsx (api) для использования в моделях/фичах и т.д.
   > Но это для меня это и так было ясно на самом деле, что папка api вряд ли на таком уровне нужна будет =)
![image](https://user-images.githubusercontent.com/42924400/109695800-6769c900-7b9d-11eb-9af3-b8377f35d8a4.png)

**Comment by azinit** [comment]
Я понимаю, когда нам нужны сущности для решения инфраструктурных проблем (тип хелперов / компонентов связанные с конкретной сущностью)

Но когда нам пригождается там шарить стейт?)

**Reply by azinit** [reply]
Сюда же:

@sergeysova @karina-drummer 

Я так и не понял почему вдруг стало удобнее называть `entities` а не `models`))
![image](https://user-images.githubusercontent.com/42924400/109697299-122eb700-7b9f-11eb-8b84-021edc61c090.png)

**Reply by azinit** [reply]
Тут прям оч напрашивается вопрос - "а для чего вы конкретно юзаете модели/сущности?"

**Reply by sergeysova** [reply]
Чтобы не путать модели в фичах и в сущностях с самими сущностями

**Reply by akaia-shadowfox** [reply]
Полагаю, что причина в том, что сущность может иметь не только shared state, но и компоненты. Модель — это только про данные, поэтому наличие компонентов в models/.../ может запутать

**Reply by azinit** [reply]
Вот в том и проблема)

Мы у себя фичи называем скорее как-то так:

- tasks-list
- task-details
- issues-map

И т.д.

И вроде даж @sergeysova что-то похожее показывал на записи

![image](https://user-images.githubusercontent.com/42924400/109700833-56bc5180-7ba3-11eb-8507-62c1d97cd3d7.png)


Однако и тут появляются модели... Вот что-бы что)

У нас если уже есть отлаженная работа с моделями (`src/entities`), зачем для фичей выделять отдельно место для них выделять

Т.е. если абстракция должна в себе хранить другую

- `features/.../features`
- `features/.../models`

- `processes/.../models`

=> Все это в моем понимании просто лишний раз намекает, что мы что-то делаем не так

А щас, когда у нас будут `features/.../models` и `entities` - как по мне будет еще больше вопросов)

У нас похожая шляпа была с фичами и процессами (но благо там разрулилось почти само собой)

Но когда у нас в проекте еще одновременно будут модели и entities... Ну хз ребят, хз))

Пока не вижу очевидных прям причин, чтобы отдельно в фиче каждой лежали модели

**Reply by azinit** [reply]
Повторюсь, у нас до этого в feature-driven обычно в фиче что-то такое лежало (и вродь нормально себя зарекомендовало, не нужно было еще отдельно плодить модели для каждой фичи)

https://github.com/feature-sliced/wiki/blob/master/FAQ.md#how-to-structure
![image](https://user-images.githubusercontent.com/42924400/109701115-b4509e00-7ba3-11eb-92da-f06974ee2349.png)

**Reply by akaia-shadowfox** [reply]
> Тут прям оч напрашивается вопрос - "а для чего вы конкретно юзаете модели/сущности?"

Я пока что разместила там i18n, notifications, skins — у всего есть как shared state, состоящий из нескольких models, так и components

**Reply by azinit** [reply]
> что сущность может иметь **не только shared state, но и компоненты. Модель — это только про данные**, поэтому наличие компонентов в models/.../ может запутать

Посмею не согласиться))

- Я как раз таки и пытался понять вашу парадигму, с чего вы вдруг решили сущности отдельно вынести - с точки зрения, что есть компоненты, которые не положишь в общий UIKit и которые связаны чисто с моделью определенной)

- Также есть и хелперы/обработчики/либы - тоже для конкретной модели

- Стейт.... Ну я вот щас вспоминаю, как некоторые ребята по работе вместо фич делили по моделям (т.е. была папка `features/users`, `features/tasks` и т.д.), но это тоже себя по итогу не зарекомендовало насколько понял


Поэтому ключевые вопросы остаются:
1. Зачем в моделях/сущнтостях стейт
2. Поч entities,а не модели
3. И все же - зачем эти модели нам нужны отдельно для фичей?

**Reply by sergeysova** [reply]
Я предполагаю, что в фичах лежит код, который несёт бизнес ценность.
Какую ценность несёт название task-list или task-details?

**Reply by azinit** [reply]
> Я предполагаю, что в фичах лежит код, который несёт бизнес ценность.

Абсолютли

> Какую ценность несёт название task-list или task-details?

То что эта фича как-то подгружает, отображает и позволяет работать с списком-задач / карточкой-задачи

И с этой (возможно слишком примитивной для тебя) точки зрения - это и есть фича)

> Она есть на странице, она комплексная, она содержит в себе только то что нужно и она автономная
![image](https://user-images.githubusercontent.com/42924400/109702081-d5fe5500-7ba4-11eb-87b6-03d212e3ae28.png)

**Reply by akaia-shadowfox** [reply]
Я понимаю это так: у фичи в модели state, который нужен только этой фичи, сама фича первична. В случае с entities первичен сам state, к которому можно получить доступ из разных мест приложения. Components же в данном случае либо предоставляют этот доступ, либо позволяют взаимодействовать со state.

**Reply by azinit** [reply]
> Я пока что разместила там i18n, notifications, skins — у всего есть как shared state, состоящий из нескольких models, так и components

А почему это не может быть отдельной фичей?

*"У меня есть фича `i18n`. У нее есть своя логика. Она несет бизнес ценность. У нее есть свой комплексный стейт, который влияет на работу. Это часть бизнес-функционала"*

То же самое про работу с нотификациями и skins (сложно угадать что это)

@sergeysova еще говорил про "Навигацию", но мне кажется это тоже можно "продать архитектуре" как фичу

И вроде должно работать

С этой же точки зрения и `auth` должна быть фичей, и возможно именно в этом месте у нас рождаются разногласия:
- Я бы сделал авторизацию фичей (со всем что надо внутри, и экспортируя самое нужное)
- А вы бы возможно еще как-то разбили по сущностям (@sergeysova , @karina-drummer )

Будет классно, если поправите где не прав))

**Reply by azinit** [reply]
> Я понимаю это так: у фичи в модели state, который нужен только этой фичи, сама фича первична. В случае с entities первичен сам state, к которому можно получить доступ из разных мест приложения. Components же в данном случае либо предоставляют этот доступ, либо позволяют взаимодействовать со state.

Вот теперь похоже понемножку начинаю понимать... **Т.е. у нас стейт сущности (user) может жить отдельно от стейта фичи (user-...)**?

**Reply by azinit** [reply]
*Отдельно - в плане, что фичи уже может не быть на странице, а в стейте значение еще есть*

**Reply by azinit** [reply]
Я прост реально пытаюсь понять, зачем стейт фичи еще как-то разбивать по моделям (которые внутри фичи)

Разве он не должен быть единым и целостным?

**Reply by azinit** [reply]
Или например, если единственное что делают модели в фиче - это хранят стейт

То тогда с тем же успехом можно просто завести папку `feature/some/store/...` и там уже хранить нужные стейты (так по мне более чисто выходит, когда заходишь в фичу: а там есть отдельно `components` (ui логика), и отдельно `store` (работа со стором)

Понятно, что инфраструктурно и другие приблудины могут добавиться... Но зачем прям выносить отдельно модели, как самостоятельные сущности... Да и на верхний уровень фичи... Оч надеюсь что смогу с вашей точки зрения понять))

**Reply by sergeysova** [reply]
Вот это "как-то работать" со списком это не бизнес-ценность

**Reply by azinit** [reply]
> Вот это "как-то работать" со списком это не бизнес-ценность

Не, ну если ты раскроешь структуру - увидешь какие там экшоны в сторе творятся, какие основные компоненты (например `Toolbar`, `Filters`, `Content`) - ты уже более подробно узнаешь что это за "как-то работает"

> UPD: Просто мы потому и инкапсулируем, потому что бизнес когда говорит "список задач" не проговаривает почти никогда дополнительно "это тот самый который умеет то-то, то-то и еще то-то" - это уже реализация внутри, которая хоть и должна быть явной и на верхнем уровне - но я чет пока не особо представляю а как бы ты делал (см. ниже)

**Reply by azinit** [reply]
@sergeysova Окей

Вот тебе надо реализовать фичу, которая рендерит список задач (или блог постов, не суть), и должна уметь:
- фильтровать их (разными способами)
- искать по определенным полям
- отображать данные в виде таблицы

Как бы ты реализовал **архитектурно**? (и называл сущности)

**Reply by sergeysova** [reply]
Значит название не отражает смысла фичи.
У бизнеса ты не увидишь фичу с таким названием

**Reply by sergeysova** [reply]
Ты сейчас перечислил детали реализации

**Reply by azinit** [reply]
Будет классно тогда, если ты назовешь, что в твоем понимании "фича которая представляет бизнес-ценность для бизнеса"

**Reply by sergeysova** [reply]
Представь себе небольшую стори, ее можно разбить на задачи не имеющие деталей реализации, это и будет фича

**Reply by sergeysova** [reply]
Представь, что ты на митинге с бизнес командой, ты хочешь рассказать о своих планах на ближайший спринт, очевидно, что рассказывать бизнесу о "task-detail" не имеет смысла, для бизнеса это деталь реализации, так как этот компонент может использоваться много где, в разных процессах, на разных страницах.
И вообще, task-detail это не фича, это компонент.

**Reply by sergeysova** [reply]
Фича должна принести пользу конечному пользователю, в идеале достаточно маленькую, чтобы можно было компонировать.
Но при этом пользу "task-detail" не может принести пользу, так как здесь нет решаемой задачи в описании, лишь есть название. Фича должна решать маленькую, но четкую задачу **пользователя**.

Можно подумать, что отображение данных это задача, но на самом деле весь фронтенд в первую очередь про отображение, эта задача решается компонентом буквально. Компоненты относящиеся к task как раз можно легко положить в entities/task.

Если говорить о фильтрации как о задаче решаемой фичей, то тут тоже вопрос, неужели фильтрация существует только фильтрация тасков? Если нет, то вот и сущность entities/filter, где будет располагаться базовый набор компонентов, хуков и фабрик упрощающих построение фильтрации.

А теперь давай вернёмся к фича фильтрации? Эта фильтрация где делается? Не можешь же ты отфильтровать одну задачу, скорее всего речь идёт о отсеивании результатов поиска, который уже может нести бизнес ценность. Но в фичу не хочется закладывать конкретную страницу с поиском и фильтрами, но хочется заложить все возможные фильтры по таскам.

Вот и выплывает название из задачи, которую решает фича: "реализация фильтрации поиска по таскам" или чуть более корректно "фильтр в поиске по таскам".
`features/task-list-search-filter`

Эта фича импортнет в себя компоненты и хуки из `entities/filter` для конфигурации обобщено вида фильтров; компоненты, хуки, логику и немного моделей из `entities/task` для отображение корректных фильтров по параметрам тасков, а также вызов правильных методов при фильтрации, если используются реестры/словари значений или кеши.

Почему в названии фичи есть search? Потому что можно придумать ещё десяток бизнесовых фич связанных с поиском, но не фильтрацией и если быть консистентный, то не придётся переименовывать существовавшие до этого фичи, ради соответствия новой функциональности, а значит не поймаем конфликты при мерже из-за переименованных директорий.

**Reply by ilyaagarkov** [reply]
>  вот и сущность entities/filter, где будет располагаться базовый набор компонентов, хуков и фабрик упрощающих построение фильтрации

чет по мне filters это скорее к libs и ui чем к entities относится

**Reply by azinit** [reply]
@ilyaagarkov вот тут как раз по-моему что-то среднее:
- функционал фильтрации это вряд ли `features/tasks-list-filter`, т.к. `list-filter` может быть и не только для тасков
   > Т.е. это вряд ли `feature` (тут больше вроде понимаю @sergeysova )
- функционал фильтрации это вряд ли набор разрозненных ~~хелперов~~ компонентов (ui) и библиотек - это скорее что-то более 
   > Т.е. это вряд ли `shared` ~ libs/ui

А потому - это скорее что-то промежуточное - т.е. именно entities (хотя сам еще не могу до конца принять именно такую трактовку и прикладное применение - но что-то  этом есть)

**Reply by azinit** [reply]
На [стриме](https://www.youtube.com/watch?v=RQBslp8dngA) по-моему похожий пример обсуждали (где-то после середины вродь)

**Reply by azinit** [reply]
UPD: Обсудили вчера еще раз с @sergeysova и @karina-drummer ситуацию по `features/.../models`

> @karina-drummer 
> Полагаю, что причина в том, что сущность может иметь не только shared state, но и компоненты. Модель — это только про данные, поэтому наличие компонентов в models/.../ может запутать



Скорее всего имелось в виду следующее:
- `features/{foo}` - конкретная фича (максимально близкая к бизнес-ценности, в зависимости от проекта и точки зрения его архитектора)
- `enitites/{bar}` - какая-либо сущность/модель из БЛ - которые помогают реализовать логику выше стоящим абстракциями (features, pages, ...)

А вот
- `features/{foo}/models` - это скорее некий `store` фичи (хотя ребята побьют меня за это определение, но пока я его трактую именно так)
   -  Т.е. это именно некая средняя абстракция, которая помогает фича общаться событиями с внешним миром (т.е. его модель или для нас редуксоидов - store)

**Reply by azinit** [reply]
> @ilyaagarkov если захочешь продолжить [дискуссию "Что есть фича" - welcome :)](https://github.com/feature-sliced/wiki/discussions/23)

**Reply by akaia-shadowfox** [reply]
> > Я понимаю это так: у фичи в модели state, который нужен только этой фичи, сама фича первична. В случае с entities первичен сам state, к которому можно получить доступ из разных мест приложения. Components же в данном случае либо предоставляют этот доступ, либо позволяют взаимодействовать со state.
> 
> 
> 
> Вот теперь похоже понемножку начинаю понимать... **Т.е. у нас стейт сущности (user) может жить отдельно от стейта фичи (user-...)**?

Да, именно, если этот стэйт должен быть доступен за рамками фичи

**Reply by sergeysova** [reply]
Обсуждали вопрос на [1 созвоне комитета](https://www.youtube.com/watch?v=RQBslp8dngA)

**Comment by azinit** [comment]
Обычно под словом темплейт, я представляю что-то родственное с `Layout`

И скорее всего это просто какая-то разметка зон, в которые можно что-то вставить
> Layout.Header, Layout.Aside, LayoutSidebar, Layout.Nav и т.д.

И насколько нужно эти темплейты пригвозживать рядом с фичами, а не в shared.... нуууу ,хизе)

Обычно в моем понимании у нас Layout используется только на странице, а уже а элементы лэйаута вставляются какие-то фичи:

```tsx
// src/pages/tasks/index.tsx
const TasksPage = () => {
  ...
  return (
      <Layout className="tasks-page">
          <Layout.Aside>
              <FeatureFirst ... />
          </Layout.Aside>
          <Layout.Main>
              <FeatureSecond ... />
          </Layout.Main>
          <Layout.Sidebar>
              {/* И даже так*/}
             <FeatureThird ...>
                 <FeatureFourth ... />
             </FeatureThird>
          </Layout.Sidebar>
      </Layout >
  )
}
```

> Мб зря набрасываюсь, просто не понял прикол в entities хранить темплейты))
![image](https://user-images.githubusercontent.com/42924400/109697385-2ecaef00-7b9f-11eb-9830-be06fb9281a0.png)

**Reply by sergeysova** [reply]
Например, у тебя есть шаблон разметки страницы для пользователя. То есть оно относится не к общему ui, а к конкретной сущности.

**Reply by azinit** [reply]
А разве этот шаблон разметки для пользователя - не будет только для страницы пользователя?)

**Reply by azinit** [reply]
Т.е. `pages/user/`

**Reply by sergeysova** [reply]
А если страниц пользователя гораздо больше, чем одна, но шаблон общий?

**Reply by azinit** [reply]
> А если страниц пользователя гораздо больше, чем одна, но шаблон общий?

Т.е.:
- шаблон (компонент) используется больше чем в одном месте
- логика нетривиальна
- шаблон используется не на одной странице пользователя (их несколько)

**Не значит ли разве все это вместе взятое, что наш шаблон страницы пользователя, уже не просто шаблон для пользователя?  А какая-то конкретная версия шаблона, не относящаяся прямо к нему**

(т.е. с тем же успехом мы можем разделить этот шаблон на более абстрактные логические куски, вынести в shared и пользоваться такой же разметкой, например, для "страницы организации")

**Reply by azinit** [reply]
И по итогу получится какой-нибудь `DetailsLayout` (условно)

Который и юзается в нескольких местах, и в том числе прямо не привязан к пользователю

Потому что с тем же накидом можно (как будто) сказать - а вот что если лэйаут для страниц пользователя используется еще для организаций? (ну вот так решили по дизайну, там схожие элементы: Основная часть, Список чего либо, Сайдбар и ...)

**Reply by sergeysova** [reply]
В этом и прикол, что шаблон пользователя позволяет сгруппировать схожие страницы с общим layout

**Reply by sergeysova** [reply]
> Не значит ли разве все это вместе взятое, что наш шаблон страницы пользователя, уже не просто шаблон для пользователя? А какая-то конкретная версия шаблона, не относящаяся прямо к нему

> И по итогу получится какой-нибудь DetailsLayout (условно)
> Который и юзается в нескольких местах, и в том числе прямо не привязан к пользователю

В идеале так и должно быть, но зачастую бывают ситуации, когда есть несколько страниц пользователя, с одинаковым шаблоном специфичным именно для пользователя (со стейтом например).

В таких случаях не зазорно его положить в `entities/user/template/*.tsx`

**Comment by azinit** [comment]
@karina-drummer 
> templates у нас в app будут? 🤔
Просто не могу представить где их ещё хранить, если речь идёт не о темплейте, который нужен только конкретной странице

**Reply by azinit** [reply]
Думаю особенно с вашим Atomic Design как-то поудобнее внутри там можно расположить
Но с @sovasergey также пришли к тому, что для некоторых случаев и в entities можно специфичные templates хранить

**Reply by azinit** [reply]
@karina-drummer 
> А если темплейт содержит логику?

Можешь пример привести?
Я под темплейтом подразумеваю просто "мета-разметку" страницы (Layout.Header, Layout.Nav, Layout.Aside, Layout.Sidebar, ...)
Понятно, что там могут быть и специфичные зоны, но прям логика какая-то.... Ну хз,  если это логика визуальная - то вродь норм
А если нужен стейт какой-то - то не совсем понимаю зач)

@karina-drummer 
> Просто если темплейт для лэйаута всего приложения, особенно если туда ещё и компонент из фичи надо подтянуть ( например, инпут поиска, который должен всегда быть в хэдэре на любой странице ) то как-то хз какие варианты ещё могут быть

**Reply by azinit** [reply]
@karina-drummer отвечая на твой вопрос - а разве у тебя этот layout - не может принимать children компоненты?

Тип то что ты описываешь - больше похоже на проблему проектирования API самого темплейта, а не методологии (т.е. ты используешь темплейт тупо как посредник, куда в специфичные зоны можно пихнуть фичи или другие компоненты)

> Вот так например
![image](https://user-images.githubusercontent.com/42924400/110446417-32df9b00-80d0-11eb-941a-9f8d91f1d593.png)

**Reply by akaia-shadowfox** [reply]
Кейс: есть ui/header, который должен быть на любой странице и features/search, инпут которого должен в этом header лежать. Логично было бы иметь template, в который подключен header с вложенным инпутом из features/search. И в этот темплейт уже обернуть app root. Но куда тогда его положить? Или есть другие варианты решения проблемы?

**Reply by akaia-shadowfox** [reply]
@martis-git тогда придётся абсолютно в каждой странице заново писать header > search

**Reply by akaia-shadowfox** [reply]
P.S. кейс не мой, однако на моём проекте с layout API тоже всё грустно, предстоит много работы

**Reply by sergeysova** [reply]
> тогда придётся абсолютно в каждой странице заново писать header > search

Конкретно этот кейс решается созданием feature с поиском, куда и можно добавить такой заголовок.

```
src/
  features/
    global-search/
      components/
        search-header.tsx
        search-panel.tsx
      templates/
        search-template.tsx
  shared/
    ui/
      templates/
        header.tsx
```

Смысл в том, что ты в фиче поиска предоставляешь компоненты для реализации поиска в том виде, в каком требует страница. В коде самой страницы можешь композировать составляющие.

**Reply by Lehakos** [reply]
А если в хедере нужны компоненты из разных фич? И такой хедер должен быть на каждой странице

**Reply by azinit** [reply]
@Lehakos я обычно такие вещи так разруливал:

1. Раз хедер используется на всех страницах - либо я заявляю об этом на уровне `pages`  (Сначала Header в разметке, и рядом Switch для роутов), либо на уровне `app`
    > Оч зависит от того - должен ли иметь доступ хедер к роуту, и в целом - о чем должен знать хедер
    > 
    > *А также от того - может ли быть так, что на определенных страницах этого хедера все же не будет*

2. Поскольку чаще всего выбирал app - то, обычно в `src/app/` и располагал компонент хедера (т.к. другим абстракциям он не нужен просто)

> [Здесь пример как я раньше делал обычно](https://github.com/ani-team/github-client/blob/dev/src/app/index.tsx#L17) - мб не самый лучший пример, но суть ясна

--- 

Но т.к. у наших товарищей (@sergeysova , @karina-drummer ) общий Layout может оччч сильно видоизменяться  от странице к страницы - думаю они получше развернут (так понимаю - выше рассуждения про это как раз)

**Reply by azinit** [reply]
Иногда такие элементы выносил как фичу (т.к. там логика тож будь здоров), и потом импортил на нужных страницах

**Reply by ilyaagarkov** [reply]
Давайте попробуем поискать место для хедера.

`app` - подходит отлично если хедер статичен, уровень app может импортить в себя любые фичи. Если хедер отличается от страницы к странице то `app` не подходит, так как страницы не могут лезть в `app`

`processes` - процесс навигации звучит неплохо, но проблемы те же самые что с `app`

`pages` - могут импортить любые фичи, но в pages у нас не отведено место под общие компоненты

`featrues` - фича навигация звучит почти не плохо(хотя уже и спорно), но фичи не  должны импортить другие фичи

все что ниже не имеет смысла описывать, так как уровень ниже не могут импортить в себя фичи.

Я вижу два варианта,

1. мы размещаем каким-то образом это все в pages, тогда никакие границы нарушены не будут. И в целом пользователи хедера это только pages 
2. мы делаем фичу навигации и нарушаем правило импорта фичи из фичи.

Мне ближе первый вариант

P.S. нужно подумать какие еще компоненты могут вызвать такие же проблемы. Помимо хедера могут быть еще сайд бары, как пример

**Reply by ilyaagarkov** [reply]
В дополнение к предыдущему

Класть навигацию в `featrues`, так же кажется не правильным, ведь фича это что-то что несет бизнес ценность, а когда у нас навигация предполагается в 100% проектах, разрабатываемых по feature-sliced, то можно сразу сказать что навигация это ни разу не про бизнес ценность, а значит ей не место в `featrues`

**Reply by azinit** [reply]
> навигация это ни разу не про бизнес ценность, а значит ей не место в featrues

Согласен, поэтому обычно стараюсь резолвить навигацию/роутинг на уровне `pages`

> Хотя как говорил @sergeysova - чаще всего не всегда так просто, и приходится резолвить на уровне тех же `processes`
